docs/specifications/4-standard-compliance-suite.md

```diff
@@ -0,0 +1,500 @@
+
+// THEN
+expect(bluePillA.pill instanceof BluePill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(bluePillB.pill instanceof BluePill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeTruthy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(redPillA.pill instanceof RedPill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(redPillB.pill instanceof RedPill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeTruthy();
+```
+
+##### Kernel Trace
+```
+```
+</details>
+
+#### Partially initialized object consumption
+When a constructor passes `this` out from **JavaScript** to the *host* app, the
+reference must be correctly identified and passed across.
+
+> :construction: The .NET Runtime does not currently honor object identity,
+> meaning that despite the same object reference is returned twice, two distinct
+> proxies exist for it in the *host* .NET app.
+
+<details><summary>Show test</summary>
+
+##### Reference Implementation
+```ts
+// GIVEN
+export abstract class PartiallyInitializedThisConsumer {
+  public abstract consumePartiallyInitializedThis(obj: ConstructorPassesThisOut): void;
+}
+export class ConstructorPassesThisOut {
+  public constructor(consumer: PartiallyInitializedThisConsumer) {
+    consumer.consumePartiallyInitializedThis(this);
+  }
+}
+
+// WHEN
+class MyConsumer extends PartiallyInitializedThisConsumer {
+  public obj?: ConstructorPassesThisOut = null;
+
+  public consumePartiallyInitializedThis(obj: ConstructorPassesThisOut) {
+    this.obj = obj;
+  }
+}
+const consumer = new MyConsumer();
+const object = new ConstructorPassesThisOut(consumer);
+
+// THEN
+expect(consumer.obj).toBe(object);
+```
+
+##### Kernel Trace
+```
+# < {"hello":"@jsii/runtime@..."}
+# > {"api":"load","name":"...","version":"...","tarball":"..."}
+# < {"ok":{"assembly":"...","types":2}}
+
+> {"api":"create","fqn":"test.PartiallyInitializedThisConsumer","args":[],"overrides":[{"method":"consumePartiallyInitializedThis"}],"interfaces":[]}
+< {"ok":{"$jsii.byref":"test.PartiallyInitializedThisConsumer@10000"}}
+> {"api":"create","fqn":"test.ConstructorPassesThisOut","args":[{"$jsii.byref":"test.PartiallyInitializedThisConsumer@10000","$jsii.interfaces":[]}],"overrides":[],"interfaces":[]}
+< {"callback":{"cbid":"jsii::callback::20000","invoke":{"objref":{"$jsii.byref":"test.PartiallyInitializedThisConsumer@10000"},"method":"consumePartiallyInitializedThis","args":[{"$jsii.byref":"test.ConstructorPassesThisOut@10001"}]}}}
+> {"complete":{"api":"complete","cbid":"jsii::callback::20000"}}
+< {"ok":{"$jsii.byref":"test.ConstructorPassesThisOut@10001"}}
+```
+</details>
+
+### Interfaces
+Tests in this section ensure correct behavior of *behavioral interfaces*.
+
+#### Host app can implement an interface "from scratch"
+It is possible for a "pure" host interface implementation to be passed across
```

Comment:
I'm only now noticing that you are planning to write descriptive notes and language-specific implementation notes with every test. So is the plan to collect a giant swath of tests, with fragments of prose surrounding them, and the aggregation of all of those is the spec? 

I would have expected to see a long-form document that talks me through the behavior of jsii in general, a document per language with implementation notes, and references out to a set of tests.

Reply:
The organization is obviously subject to change as we iterate and discover what is pragmatic or not.

The language-specific notes at this point are mostly tokens of known existing non-compliance of runtimes that we however already offer as "Generally Available". They help a reader not get confused in a way akin to:

>The specification says X but I can observe X', what the eff?

I am *extremely* favorable to building up a document that details the normative behavior in a clear & concise way, and then replace most of the prose attached to the tests with references to specific points in that document (or/and vice-versa). I believe the test suite will help formalize this document better.